lib/api/apiUtils/object/corsResponse.js

```diff
@@ -0,0 +1,130 @@
+/** _matchesValue - compare two values to determine if they match
+* @param {string} allowedValue - an allowed value in a CORS rule;
+    may contain wildcards
+* @param {string} value - value from CORS request
+* @return {boolean} - true/false
+*/
+function _matchesValue(allowedValue, value) {
+    const wildcardIndex = allowedValue.indexOf('*');
+    // If no wildcards, simply return whether strings are equal
+    if (wildcardIndex === -1) {
+        return allowedValue === value;
+    }
+    // Otherwise make sure substrings match expected substrings before
+    // and after the wildcard
+    const beginValue = allowedValue.substring(0, wildcardIndex);
+    const endValue = allowedValue.substring(wildcardIndex + 1);
+    return (value.startsWith(beginValue) && value.endsWith(endValue));
+}
+
+/** _matchesOneOf - check if a value matches any values of a CORS rule array
+* @param {string[]} allowedValues - array of values allowed in CORS rule
+* @param {string} value - value from CORS request
+* @return {boolean} - true/false
+*/
+function _matchesOneOf(allowedValues, value) {
+    return allowedValues.some(allowedValue =>
+        _matchesValue(allowedValue, value));
+}
+
+/** _headersMatchRule - check if headers match AllowedHeaders of rule
+* @param {string[]} headers - Access-Control-Allow-Headers values or any headers
+*   that are sent in a CORS request
+* @param {string[]} ruleAllowedHeaders - AllowedHeaders of a CORS rule
+* @return {boolean} - true/false
+*/
+function _headersMatchRule(headers, ruleAllowedHeaders) {
+    if (!ruleAllowedHeaders) {
+        return false;
+    }
+    // AllowedHeaders may have been stored with uppercase letters
+    // during putBucketCors; ignore case when searching for match
+    const allowedHeaders = ruleAllowedHeaders.map(allowedHeader =>
```

Comment:
This is a good point, so I gave it a shot in this commit: https://github.com/scality/S3/pull/518/commits/c607e660ba4c65d353282b31b4297106f9d18a55

Note: I was also using that function to check if the origin matched any of the AllowedOrigin's as well since the code was the same. But I created the function mainly so that `_headersMatchRule` would be more readable than nesting a .some() condition in an .every(). Since we are making the `_matchesOneOf` function specific to the allowed headers, I substituted the raw code back for the allowed origins check.